File: spicedb-csharp/examples/LookupResources/LookupResourcesTest.cs
using Xunit;
// Example LookupResources demonstrates finding resources a subject can access.

using SpiceDB.Client;
using static SpiceDB.Client.Consistency;

namespace LookupResources;

public class LookupResourcesTest
{
    private const string Schema = """
        definition user {}

        definition document {
            relation viewer: user
            relation editor: user
            permission view = viewer + editor
            permission edit = editor
        }
        """;

    [Fact]
    public async Task LookupResources_FindsAccessibleDocuments()
    {
        await using var client = SpiceDBClient.CreatePlaintext(SpiceDBTestServer.Endpoint, SpiceDBTestServer.Token);

        await client.WriteSchemaAsync(Schema);

        var txn = new Transaction();
        txn.Touch(Relationship.FromTriple("document", "firstdoc", "viewer", "user", "alice"));
        txn.Touch(Relationship.FromTriple("document", "seconddoc", "editor", "user", "alice"));
        await client.WriteAsync(txn);

        // Lookup resources alice can view. Each result is a native
        // LookupResource carrying the resource ID plus permissionship — check
        // Permissionship before treating a result as a full grant, since
        // ConditionalPermission results depend on caveat context that was not
        // fully evaluated by the server.
        var resourceIDs = new HashSet<string>();
        await foreach (var result in client.LookupResourcesAsync(
            Full(), "document", "view", "user", "alice"))
        {
            Assert.Equal(Permissionship.HasPermission, result.Permissionship);
            resourceIDs.Add(result.ResourceID);
        }

        Assert.Contains("firstdoc", resourceIDs);
        Assert.Contains("seconddoc", resourceIDs);
    }

    [Fact]
    public async Task LookupResources_WithDebug_StillReturnsResults()
    {
        await using var client = SpiceDBClient.CreatePlaintext(SpiceDBTestServer.Endpoint, SpiceDBTestServer.Token);

        await client.WriteSchemaAsync(Schema);

        var txn = new Transaction();
        txn.Touch(Relationship.FromTriple("document", "thirddoc", "viewer", "user", "alice"));
        await client.WriteAsync(txn);

        // withDebug asks the server to attach debug info (currently limited
        // to maximum-recursion-depth errors) to the error details of a
        // failed call. It must not change the results of a call that
        // succeeds.
        var resourceIDs = new HashSet<string>();
        await foreach (var result in client.LookupResourcesAsync(
            Full(), "document", "view", "user", "alice", withDebug: true))
        {
            resourceIDs.Add(result.ResourceID);
        }

        Assert.Contains("thirddoc", resourceIDs);
    }
}

File: spicedb-java/examples/src/test/java/com/authzed/spicedb/examples/LookupResourcesTest.java
package com.authzed.spicedb.examples;

import static com.authzed.spicedb.Consistency.*;
import static org.assertj.core.api.Assertions.*;

import com.authzed.spicedb.Filter;
import com.authzed.spicedb.LookupResult;
import com.authzed.spicedb.Relationship;
import com.authzed.spicedb.Transaction;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/**
 * Demonstrates finding resources a subject can access using {@link
 * com.authzed.spicedb.SpiceDBClient#lookupResources}.
 */
class LookupResourcesTest extends SpiceDBIntegrationTest {

  @BeforeEach
  void setUp() {
    client.deleteRelationships(Filter.of("document"));

    var txn = new Transaction();
    txn.touch(Relationship.of("document", "firstdoc", "viewer", "user", "alice"));
    txn.touch(Relationship.of("document", "seconddoc", "editor", "user", "alice"));
    txn.touch(Relationship.of("document", "thirddoc", "owner", "user", "bob"));
    client.write(txn);
  }

  @Test
  void alice_can_view_two_documents() {
    List<LookupResult.LookupResource> results;
    try (var stream = client.lookupResources(full(), "document", "view", "user", "alice")) {
      results = stream.toList();
    }

    // Every result carries a permissionship — callers must check it before treating a
    // CONDITIONAL_PERMISSION result as a full grant (see wildcard/caveat-aware examples).
    //
    // Results also aren't guaranteed unique: the same resourceId can be yielded more than
    // once (e.g. via caveated/conditional results), possibly with a different permissionship
    // each time. A caller that needs uniqueness must deduplicate by resourceId itself.
    assertThat(results)
        .allSatisfy(
            r ->
                assertThat(r.permissionship())
                    .isEqualTo(LookupResult.Permissionship.HAS_PERMISSION));
    List<String> resourceIDs =
        results.stream().map(LookupResult.LookupResource::resourceId).toList();
    assertThat(resourceIDs).containsExactlyInAnyOrder("firstdoc", "seconddoc");
  }

  @Test
  void alice_can_edit_only_seconddoc() {
    List<String> resourceIDs;
    try (var stream = client.lookupResources(full(), "document", "edit", "user", "alice")) {
      resourceIDs = stream.map(LookupResult.LookupResource::resourceId).toList();
    }

    assertThat(resourceIDs).containsExactly("seconddoc");
  }

  @Test
  void bob_can_delete_thirddoc() {
    List<String> resourceIDs;
    try (var stream = client.lookupResources(full(), "document", "delete", "user", "bob")) {
      resourceIDs = stream.map(LookupResult.LookupResource::resourceId).toList();
    }

    assertThat(resourceIDs).containsExactly("thirddoc");
  }
}
